firebase-test-lab/trigger_ftl_tests.py

```diff
@@ -185,67 +186,95 @@ def _run_test_on_ftl(FLAGS):
-  # Test outcome pattern 2:
-  # OUTCOME: Passed
-  # TEST_AXIS_VALUE: redfin-30-en-portrait
-  # TEST_DETAILS: --
-  outcome_device_search = re.findall(r'OUTCOME:(.*?)\nTEST_AXIS_VALUE:(.*?)\nTEST_DETAILS:', result_log, re.MULTILINE | re.DOTALL)
-  for o_d in outcome_device_search:
-    outcome_device.append({"device_axis": o_d[1].strip(), "outcome": o_d[0].strip()})
-
-  while result.poll() is None:
-    time.sleep(1) # Process hasn't exited yet, let's wait some
-  # https://firebase.google.com/docs/test-lab/android/command-line#script_exit_codes
-  logging.info("Test done: %s\nReturned code: %s\n%s", ftl_cmd, result.returncode, result_log)
+    raw_result_link = ""
+    raw_result_link_search = re.search(r'Raw results will be stored in your GCS bucket at \[(.*?)\]', result_log, re.MULTILINE | re.DOTALL)
+    if raw_result_link_search:
+      raw_result_link = raw_result_link_search.group(1)
+    
+    outcome_device = []
+    # Test outcome pattern 1:
+    # ┌─────────┬──────────────────────────────┬─────────────────────┐
+    # │ OUTCOME │       TEST_AXIS_VALUE        │     TEST_DETAILS    │
+    # ├─────────┼──────────────────────────────┼─────────────────────┤
+    # │ Failed  │ iphone13pro-15.2-en-portrait │ 1 test cases failed │
+    # └─────────┴──────────────────────────────┴─────────────────────┘
+    outcome_device_search = re.findall(r'│(.*?)│(.*?)│(.*?)│', result_log, re.MULTILINE | re.DOTALL)
+    for o_d in outcome_device_search:
+      if 'OUTCOME' in o_d[0]: # skip the table title
+        continue
+      outcome_device.append({"device_axis": o_d[1].strip(), "outcome": o_d[0].strip()})
+    # Test outcome pattern 2:
+    # OUTCOME: Passed
+    # TEST_AXIS_VALUE: redfin-30-en-portrait
+    # TEST_DETAILS: --
+    outcome_device_search = re.findall(r'OUTCOME:(.*?)\nTEST_AXIS_VALUE:(.*?)\nTEST_DETAILS:', result_log, re.MULTILINE | re.DOTALL)
+    for o_d in outcome_device_search:
+      outcome_device.append({"device_axis": o_d[1].strip(), "outcome": o_d[0].strip()})
+    
+    return {
+      "attempt": attempt_num,
+      "cmd": ftl_cmd,
+      "return_code": result.returncode,
+      "testapp_path": testapp_path,
+      "test_type": FLAGS.test_type,
+      "ftl_link": ftl_link,
+      "raw_result_link":  raw_result_link,
+      "devices": outcome_device
+    }
+
+
+def _validate_results(FLAGS, test_summary):
+  if test_summary.get("return_code") != 0:
+    return False
   
-  test_summary =  {
-    "cmd": ftl_cmd,
-    "return_code": result.returncode,
-    "testapp_path": testapp_path,
-    "test_type": FLAGS.test_type,
-    "ftl_link": ftl_link,
-    "raw_result_link":  raw_result_link,
-    "devices": outcome_device
-  }
-  tests_result.get('apps').append(test_summary)
+  if FLAGS.test_type == GAMELOOP and FLAGS.validator:
+    try:
+      module = os.path.splitext(FLAGS.validator)[0]
+      validator = imp.load_source(module, FLAGS.validator)
+      return validator.validate(test_summary)
```

Comment:
Users will need the class/interface definition for validator and test_summary in order to provide their implementation. Do you want to clarify that?

Reply:
It is exposed as an input option in the action yaml, isn't it? Still it's up to you.

Reply:
I think it's fine, I also added "[Experimental]" to the input description.
If you have any ideas on how to hide this input from action yaml, please let me know.